File: Go/Ejercicio 3/gestion_pila_cartas.go
package main

import (
	"math/rand"
	"time"
)

type GestionPilaCartas struct {
	pila []string
}

func NuevaGestionPilaCartas() *GestionPilaCartas {
	cartas := []string{"A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"}
	baraja := []string{}
	for i := 0; i < 4; i++ {
		baraja = append(baraja, cartas...)
	}
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(baraja), func(i, j int) { baraja[i], baraja[j] = baraja[j], baraja[i] })
	return &GestionPilaCartas{pila: baraja}
}

func (g *GestionPilaCartas) SacarCarta() (string, bool) {
	if len(g.pila) == 0 {
		return "", false
	}
	carta := g.pila[len(g.pila)-1]
	g.pila = g.pila[:len(g.pila)-1]
	return carta, true
}

func (g *GestionPilaCartas) CartasRestantes() int {
	return len(g.pila)
}

File: Typescript/Ejercicio 3/gestion_pila_cartas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GestionPilaCartas = void 0;
var GestionPilaCartas = /** @class */ (function () {
    function GestionPilaCartas() {
        this.pila = this.crearBaraja();
    }
    GestionPilaCartas.prototype.crearBaraja = function () {
        var cartas = [
            "A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"
        ];
        var baraja = [];
        for (var i = 0; i < 4; i++) {
            baraja.push.apply(baraja, cartas);
        }
        return this.shuffle(baraja);
    };
    GestionPilaCartas.prototype.shuffle = function (array) {
        return array.sort(function () { return Math.random() - 0.5; });
    };
    GestionPilaCartas.prototype.sacarCarta = function () {
        return this.pila.pop();
    };
    GestionPilaCartas.prototype.cartasRestantes = function () {
        return this.pila.length;
    };
    return GestionPilaCartas;
}());
exports.GestionPilaCartas = GestionPilaCartas;
